### PR TITLE
Change protection algorithm (alternative)

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #12 #15

This is alternative PR #16 .

Added `CsrfTokenFetcherInterface` with two implementations: with state and stateless.

Use package:
- Add to app middleware `CsrfMiddleware`
- Implement `CsrfTokenGeneratorInterface` (default `RandomCsrfTokenGenerator`)
- Implement `CsrfTokenFetcherInterface` (`StateCsrfTokenFetcher` or `StatelessCsrfTokenFetcher`, default with state)
- If use implementation of fetcher with state, then need implement `CsrfTokenStorageInterface` (default `SessionCsrfTokenStorage`)
- Use method `getValue` of class `CsrfToken` for get token value for send to front.

Review this concept, if ok then I add phpdoc and tests. Seems this better than #16 .

After merge need also merge PRs:
- https://github.com/yiisoft/yii-view/pull/22

### CODE REVIEW

**Comment by samdark** [comment]
@vjik need an implementation of any stateless CSRF validation to verify that interface is correct. Other than that concept looks alright.

**Comment by vjik** [comment]
> @vjik need an implementation of any stateless CSRF validation to verify that interface is correct. Other than that concept looks alright.

Moved token generator to `Stateful` namespace, improve stateless logic and make `StatelessCsrfToken`.

File: src/Stateless/StatelessCsrfToken.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return hash_equals(
+            $hash,
+            $this->generateHash($salt, $expiration)
+        );
+    }
+
+    private function generateHash(string $salt, ?int $expiration): string
+    {
+        return StringHelper::base64UrlEncode(
+            hash_hmac(
```
Comment (samdark) [inline src/Stateless/StatelessCsrfToken.php:92]: Can `Mac` from yiisoft/security be used in this case?

File: tests/CsrfMiddlewareTest.php [hunk 2]
```diff
@@ -165,18 +167,25 @@ private function createStorageMock(string $returnToken): CsrfTokenStorageInterfa
...
 
         $mock
-            ->expects($this->once())
             ->method('get')
             ->willReturn($returnToken);
 
         return $mock;
     }
 
+    private function createCsrfMiddleware(bool $autoGenerateToken): CsrfMiddleware
```
Comment (samdark) [inline tests/CsrfMiddlewareTest.php:176]: The test class could be made abstract with this and the following method marked as abstract protected. Then two test cases could be created, one for stateful token and another for stateless token.

File: tests/MaskedCsrfTokenTest.php [hunk 3]
```diff
@@ -36,6 +28,8 @@ private function createCsrfToken(string $token = null): CsrfToken
                 ->method('get')
                 ->willReturn($token);
         }
-        return new CsrfToken($mock);
+        return new MaskedCsrfToken(
```
Comment (samdark) [inline tests/MaskedCsrfTokenTest.php:31]: Same here. These tests should, ideally, be ran against both stateful and stateless tokens.

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/MaskedCsrfToken.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        $this->token = $token;
+    }
+
+    /**
+     * @return string The currently valid token string with the mask applied.
+     */
+    public function getValue(): string
+    {
+        return TokenMask::apply(
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/MaskedCsrfToken.php:29]: Made `MaskedCsrfToken` as decorator of `CsrfTokenInterface`.

File: config/web.php [hunk 5]
```diff
@@ -2,9 +2,22 @@
...
+use Yiisoft\Csrf\CsrfTokenInterface;
+use Yiisoft\Csrf\Stateful\CsrfTokenStorageInterface;
+use Yiisoft\Csrf\Stateful\RandomCsrfTokenGenerator;
+use Yiisoft\Csrf\Stateful\SessionCsrfTokenStorage;
+use Yiisoft\Csrf\Stateful\StatefulCsrfToken;
+use Yiisoft\Csrf\Stateless\CsrfTokenIdentificationInterface;
+use Yiisoft\Csrf\Stateless\SessionCsrfTokenIdentification;
 
 return [
+    CsrfTokenInterface::class => StatefulCsrfToken::class,
```
[the remark this answers is not in the document]
Comment (vjik) [reply config/web.php]: Done. But need verify...

File: src/Stateless/StatelessCsrfToken.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        $this->lifetime = $lifetime;
+    }
+
+    public function getValue(): string
+    {
+        $salt = Random::string(8);
+        $expiration = $this->lifetime === null ? null : (time() + $this->lifetime);
+
+        return StringHelper::base64UrlEncode(
+            $this->generateHash($salt, $expiration) . '~' . $salt . '~' . (string)$expiration
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Stateless/StatelessCsrfToken.php]: Removed.

File: src/Stateless/StatelessCsrfToken.php (hunk 1 above)
Comment (vjik) [reply src/Stateless/StatelessCsrfToken.php]: Done

File: tests/CsrfMiddlewareTest.php (hunk 2 above)
Comment (vjik) [reply tests/CsrfMiddlewareTest.php:175]: Done

File: tests/MaskedCsrfTokenTest.php (hunk 3 above)
Comment (vjik) [reply tests/MaskedCsrfTokenTest.php:31]: Done

**Comment by vjik** [comment]
Tests in yii-demo pass.

With this PR need also https://github.com/yiisoft/yii-view/pull/22

**Review by samdark — APPROVED** [review]

File: src/Stateless/CsrfTokenIdentificationInterface.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Csrf\Stateless;
+
+interface CsrfTokenIdentificationInterface
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/Stateless/CsrfTokenIdentificationInterface.php]: No, we can't. The meaning is different.

**Comment by samdark** [comment]
Awesome one!

**Pull request merged**